More (much more) I/O content (#740)
root-project/web: manual/io/index.md

---
title: I/O Concepts
layout: single
sidebar:
  nav: "manual"
toc: true
toc_sticky: true
---

This page explains the mechanisms and concepts behind ROOT's I/O facilities, i.e. how ROOT converts your objects into a stream of bytes and back.
It assumes that you have read the introduction on → [ROOT files]({{ '/manual/root_files/' | relative_url }}), → [I/O of custom classes]({{ '/manual/io_custom_classes/' | relative_url }}), and → [Trees]({{ '/manual/trees/' | relative_url }}).

## ROOT files, directories, and keys

ROOT stores data in ROOT files ({% include ref class="TFile" %}), see → [ROOT files]({{ '/manual/root_files/' | relative_url }}).
It does so in a machine independent format.

Similar to a file system directory, a `TFile` can contain directories ({% include ref class="TDirectory" %}) and objects, accessible through the directory's _keys_ ({% include ref class="TKey" %})).
A `TFile` is a directory itself: it inherits from `TDirectory`.

> Directories can be "entered" by [`TDirectory::cd()`](https://root.cern/doc/master/classTDirectory.html#af42d9b50db9b4575b13ea265ad8995fb); created by [`TDirectory::mkdir()`](https://root.cern/doc/master/classTDirectory.html#aad2cf362d9574db6538804afa9794827); and removed by [`TDirectory::rmdir()`](https://root.cern/doc/master/classTDirectory.html#a528126c0a528d7f7a5888c7ee74ae41d).

Reviewer: What is the process to maintain the commit information that is embedded here? (I.e. if TDirectory.h changes, how do we remember to update this commit number)?

Author: the CI at `root-project/web` complains next time anyone tries to modified the website

root [2] gDirectory->pwd()

my.root:/
root [3] hist->Draw() // for `hist` from `my.root`
{% endhighlight %}

`Rint:/` corresponds to "ROOT's in-memory" directory, the default directory during startup.

### Adding and removing objects from a ROOT file

When writing an object to a ROOT file, ROOT creates a directory "entry" ({% include ref class="TKey" %}) representing the object, which mainly consists of a name and the object's persistent data.

> You can think of a `TFile` as a collection of `TKeys`, possibly inside nested `TDirectory`-s.

Reviewer 2: `TDirectory`-s : the dash looks out of place.

Author: it's because `TDirectories` is weird and `TDirectory`s is wrong.

I'll go with `TDirectories` as the lesser evil.

The name of the `TKey` can be either explicitly stated when writing, or it can be determined from `TObject::GetName()` for classes inheriting from `TObject`.

_**Example**_

{% highlight C++ %}
void WriteHist(TH1* hist) {
  gDirectory->WriteObject(hist, "explicitName");

>   // OR:

Reviewer 2: ```suggestion
  // OR (less preferred):
```

Author: I don't think so, why?

Reviewer 2: For style reasons I believe, but it's not necessary

>   hist->Write(); // key named `TH1::GetName()` inside `gDirectory`

Reviewer 2: ```suggestion
  hist->Write(); // the key will be named as the result of `hist->GetName()` inside `gDirectory`
```

}
{% endhighlight %}

An object read into memory is independent from the object on disk.
Changes of the in-memory object are not propagated to disk.
Instead, a new version of the object needs to be saved, for instance passing `"overwrite"` as option to [ `TDirectory::WriteObject()`](https://root.cern/doc/master/classTDirectory.html#a67b115afae97366254dfd44a7f46f66f), see the documentation of [`TDirectoryFile::WriteTObject()`](https://root.cern/doc/master/classTDirectoryFile.html#ae1bb32dcbb69de7f06a3b5de9d22e852).

Removing an object from a `TDirectory` using [`TDirectory::Delete()`](https://root.cern/doc/master/classTDirectory.html#a09e51150c19fd5a0aa9278a4158202ca) will generally not free the corresponding disk space.
Instead, the storage occupied by the deleted object will be made available (as a [`TFree`](https://root.cern/doc/master/classTFree.html) for subsequent objects to be written to this file.

> Use → [`hadd`]({{ '/manual/root_files/#merging-root-files-with-hadd' | relative_url }}) to defragment a ROOT file by rewriting it.

Reviewer 2: It is clear that `Delete` does not delete the object on disk. It is not clear how to delete the object on disk. Maybe it's not the point of the sentence though

Author: Javi's understanding is correct, and it should _not_ be clear that `Delete` does not delete the object on disk, because it does :smile: it just does not free the corresponding disk space. Should we clarify somewhere, @vepadulano ?

Reviewer 2: I think I got distracted by some issue that I remember in the past where `Delete`ing an object from a TFile would not really delete it. As in, running a program that just deletes an object (and does not do anything else with the file, thus it does not fill the space that was previously occupied by that object) and then running another program that lists the content of the same file, would show the previously deleted object. Maybe I just misrember, I haven't tried to reproduce this myself again

Author: please open a github issue if that's the case :smiley:

Reviewer 2: Now I remember what was my source of confusion:
```cpp
#include <TFile.h>
#include <TH1.h>

void write(){
    TFile f{"myfile.root", "recreate"};
    TH1F h1{"h1","h1",10,0,10};
    TH1F h2{"h2","h2",10,0,10};
    f.WriteObject(&h1,"");
    f.WriteObject(&h2,"");
}

void read_and_delete(){
    TFile f{"myfile.root","update"};
    f.Delete("h1"); // change to  f.Delete("h1;1") to actually remove the object from the TFile
}

void read_and_list(){
    TFile f{"myfile.root","read"};
    f.ls();
}

int main(){
    write();
    read_and_delete();
    read_and_list();
}
```

In the `read_and_delete` function above, the "h1" histogram is only actually deleted from the TFile if the full namecycle is indicated `h1;1`. If only the name of the histogram is indicated, the `Delete` function deletes it from memory (i.e. for the lifetime of the current open file), but then closing and reopening the file again shows both histograms still there.

Reviewer 2: I'm thinking either clearer documentation in `TDirectory::Delete` or a concrete example here in the manual, or both things

Author: It's documented at https://root.cern/doc/master/classTDirectoryFile.html#a24747f020c3276e8a7d2e6c91e5e3f76

Author: (and also https://root.cern/doc/master/classTDirectory.html#a09e51150c19fd5a0aa9278a4158202ca )

Author: (i'm not saying that the behavior is intuitive, but I don't think there is more to be done here in the manual besides linking to those docs)

Reviewer 2: Yes, fair enough :+1:

### Iterating over a directory's content

`TFile` gives access to the list of keys through its base class, [TDirectoryFile::GetListOfKeys()](https://root.cern/doc/master/classTDirectoryFile.html#a8599cbd2c3fd8004e78f1f40be6771d2).

_**Example**_

Get the list of keys from the demo.root file and print them.

{% highlight C++ %}
root[] TFile f("demo.root")
root[] f.GetListOfKeys()->Print()

TKey Name = h0,     Title = histo    nr:0,     Cycle = 1
TKey Name = h1,     Title = histo    nr:1,     Cycle = 1
...
{% endhighlight %}

For instance if you do not know an object's name upfront, it can be useful to iterate through all of a directory's entries.

_**Example**_

{% highlight C++ %}
TFile f("demo.root");
// f.GetListOfKeys() contains `TObject*`, but we know they are really `TKey*`

> for (TKey *key: ROOT::RangeStaticCast<TKey*>(*f.GetListOfKeys())) {

Reviewer 2: ```suggestion
for (auto key: ROOT::RangeStaticCast<TKey*>(*f.GetListOfKeys())) {
```

Author: we actually discussed this and decided to go with explicit types (e.g. here to make it clear that `RangeStaticCast<TKey*>` returns an iterator over `TKey *`, and not `TKey&` for instance

  std::cout << "key: " << key->GetName()
    << " points to an object of class: " << key->GetClassName() << '\n';
}
{% endhighlight %}

> The output is of such an `iterate.C` ROOT macro could be:

Reviewer 3: ```suggestion
The output of such an `iterate.C` ROOT macro could be:
```

## ROOT's C++ object serialization: from memory to disk and back

Writing an object to file means writing the current values of the object's data members.
This is done with the help of what ROOT calls a _streamer_:
each serializable class has such a streamer which converts all data members into a buffer of raw bytes (see [TBuffer](https://root.cern/doc/master/classTBuffer.html)).

Variables of composite data types such as classes, structures, and arrays can be decomposed into simple types such as longs, shorts, floats, and chars.
These values are then written out in a machine-independent representation.

This happens recursively:
a second streamer will be invoked for a member of class type, to stream that class's members.
A similar recursion happens for all base classes.
In the end, the buffer contains all the simple data members of all the classes that make up that particular object.

At runtime, ROOT needs to determine which streamer function to call for a given object.
Using the `ClassDef` macro inside the class definition makes this operation more performant:
without `ClassDef`, the object's dynamic type has to be looked up to determine which streamer to invoke.

> See also → [The `ClassDef` Macro](#the-classdef-macro)).

Reviewer 3: ```suggestion
See also → [The `ClassDef` Macro]({{ '/manual/io_custom_classes/#the-classdef-macro' | relative_url }})).
```

Data members of certain types need special treatment, for instance pointers and references.
This will be explained below.

The methods of a class are not written to the ROOT file, it contains only the persistent data members.

See also → [Restrictions on types ROOT I/O can handle]({{ '/manual/io_custom_classes/#restrictions-on-types-root-io-can-handle' | relative_url }}).

### Generating streamers

Streamers are C++ functions that are usually created as part of a class's dictionary, see → [I/O of custom classes]({{ 'manual/io_custom_classes/#generating-dictionaries' | relative_url }}).

> `rootcling` parses the class definition and determines how to stream the object in an optimal way, and which streamers need to be invoked for base classes and members.

Reviewer 2: First mention of `rootcling` with no links or a few words to explain. Might be that we expect readers to know/understand what this is already

Author: at the beginning of this document we say it "assumes that you have read the introduction on → [ROOT files]({{ '/manual/root_files/' | relative_url }}), → [I/O of custom classes]({{ '/manual/io_custom_classes/' | relative_url }}), and → [Trees]({{ '/manual/trees/' | relative_url }})".

In particular "I/O of custom classes" talks about rootcling

Author: I added a link to the section @jalopezg-r00t suggested, I don't think we have anything better

### Excluding data members from I/O

To prevent a data member from being written to the file, insert a `!` as the first character in a single-line comment (`//`) following the declaration of the member.

> To accommodate doxygen-style documentation, this annotation can also be written as `///!`. 

Reviewer: ```suggestion
To accommodate doxygen-style documentation, this annotation can also be written as `///<!`. 
```

Author: Thanks, does this mean that `///->` should be `///<->`?

Reviewer: Looks very weird but I guess so (@couet @Axel-Naumann)?

Author: @couet confirmed it's `///<->`, fixed

_**Example**_

{% highlight C++ %}
class Event {
   TDirectory *fTransient; //! This won't be stored
   Float_t fPt;            // This will be stored
   int fMemberWithDoc;     ///<! The member documentation
};
{% endhighlight %}

### Marking pointers as never null

For a small performance benefit, pointer data members can be marked as always be pointing to valid memory (never being null):
this is done with the annotation `//->` or `///->`.

A pointer marked as such must not point back to the current object, not even indirectly.

_**Example**_

{% highlight C++ %}
TClonesArray *fTracks;
TH1F *fH; //->
{% endhighlight %}

> Pointers`fH` is marked as never null: ROOT will be able to perform additional optimizations.

Reviewer 2: ```suggestion
The pointer data member `fH` is marked as never null: ROOT will be able to perform additional optimizations.
```

`fTracks`, instead, will always be checked for `nullptr`.

### Array data members of fixed and variable size

ROOT supports I/O of fixed size arrays out of the box. For variable-size arrays, a special comment syntax is available to specify the name of a data member that holds the size of the array. Here is an example:

_**Example**_

{% highlight C++ %}
struct Event {
   char fType[20]; // fixed-size array, all good
   int fNVertex;
   float *fVertices; //[fNVertex]
};
{% endhighlight %}

The comment `//[fNVertex]` tells ROOT that the length of the array is stored in the corresponding variable. In general, the syntax is:

{% highlight C++ %}
TYPE *MEMBER; //[LENGTH]
{% endhighlight %}

`LENGTH` must be the name of a data member that is defined _before_ the array member, or in a base class.

> **Note**

> > Pointer to simple types (e.g. `float*`, `int*`) are assumed to be variable-size arrays. 

Reviewer 2: ```suggestion
> Pointers to simple types (e.g. `float*`, `int*`) are assumed to be variable-size arrays. 
```

If you know that a certain data member will always have to be read as a full object, it can be more performant to prevent its splitting.
To do so, add `//||` as an annotation to its declaration:

_**Example**_

{% highlight C++ %}
EventHeader fEvtHdr; //|| This data member will always be read as a whole
{% endhighlight %}

### Double32_t: storing doubles with single precision

Some values have inherent reduced precision, yet benefit from double precision arithmetics.
The type alias `Double32_t` represents a value that has double precision in memory, but it is stored with lower, adjustable precision.

The actual size on disk (before compression) is determined by the parameter next to the data element declaration:

_**Example**_

{% highlight C++ %}
Double32_t m_data; //[min,max<,nbits>]
{% endhighlight %}

If the comment is absent or does not contain `min`, `max`, `nbits`, the member is saved with single precision.

The `min` and `max` values themselves, if present, are saved with 32-bits precision. `min` and `max` can be either a floating point number or one of the following trivial mathematical expressions: `pi`, `2*pi`, `pi/2`, `pi/4`.

> If `nbits` is present, the member is saved with nbits bits precision. For more details, see [this tutorial](https://root.cern/doc/master/double32_8C.html).

Reviewer 3: ```suggestion
If `nbits` is present, the member is saved with `nbits`-bits precision. For more details, see [this tutorial](https://root.cern/doc/master/double32_8C.html).
```

A common use case is as a post-read hook, for instance for the registration of a read object with other objects.

_**Example**_

You need to tell `rootcling` not to build a streamer, by ending the `#pragma` statement with a `-`, e.g.

{% highlight C++ %}
#pragma link C++ class Event-;
{% endhighlight %}

The following is an example of a customized `Streamer` function for `Event`. It takes a {% include ref class="TBuffer" %} as a parameter, and first checks to see if this is a case of reading or writing the buffer.

_**Example**_

{% highlight C++ %}
void Event::Streamer(TBuffer &buf)
{
   if (buf.IsReading()) {
      Event::Class()->ReadBuffer(buf, this);
      fReadTime = time(); // store the time of reading
   } else {
      Event::Class()->WriteBuffer(buf, this);
   }
}
{% endhighlight %}

> **Note**
> A class with a custom streamer cannot be split, and its members cannot be stored member-wise.

### Disable storage of TObject data members in derived classes

Types do not have to inherit from {% include ref class="TObject" %} for ROOT to be able to read/write them: the presence of a dictionary is sufficient.

> Classes that do inherit from `TObject` can exclude TObject's data members from their I/O invoking `MyClass->Class::IgnoreObjectStreamer()`.

Reviewer 2: Where should `IgnoreObjectStreamer` be invoked? Some place in the class definition? Or in the user application that is writing objects of that class to a file?

Author: @pcanal @Axel-Naumann is the following correct?
```suggestion
Classes that do inherit from `TObject` can exclude TObject's data members from their I/O invoking `MyClass->Class::IgnoreObjectStreamer()` before any object of type `MyClass` is written to a ROOT file.
```

Reviewer: > before any object of type `MyClass` is written to a ROOT file.

That is correct.

Author: Added!

> Classes that do inherit from `TObject` can exclude TObject's data members from their I/O invoking `MyClass->Class::IgnoreObjectStreamer()`.

Reviewer 3: I think this should be:
```suggestion
Classes that do inherit from `TObject` can exclude TObject's data members from their I/O invoking `MyClass->Class()->IgnoreTObjectStreamer()`.
```

Author: You are right, and the instance should not be called like a type:

```suggestion
Classes that do inherit from `TObject` can exclude TObject's data members from their I/O invoking `myObject->Class()->IgnoreObjectStreamer()`.
```

- Change from variable-size array and size (such as `float *fArray; //[fSize]` and `int fSize`) to `std::vector` (such as `std::vector<float> fArray;`).
- Change between STL collection types, from / to `std::vector`, `std::queue`, `std::deque`, `std::list`, `std::forward_list`, `std::set`, `std::multiset`, `std::unordered_set`, `std::unordered_multiset`, `std::valarray`, `std::bitset`.
- Change of STL associative containers, from / to `std::map`, `std::unordered_map`, `std::multimap`, `std::unordered_map`, `std::unordered_multimap` `std::vector<std::pair<key,value>>`.

All transformations above are applied transparently with no intervention required on the part of user: ROOT will automatically recognize these cases and apply the relevant rules.

_**Example**_

Here is an example of the class layout changes that automatic schema evolution supports:

{% highlight C++ %}
// A class that was written as:
class MyClass : MyBaseClass {
   int a;
   float b;
   MyOtherClass *c;
   Pointee *d;
   std::list<Thing> e;
};

// ...can be read as:
class MyClass : MyOtherClass { // base class changed
   double b; // type changed
   int a; // order changed

>    char d; // new data member added

Reviewer 3: `d` name collides with `d` member declared in line 395.
```suggestion
   char chr; // new data member added
```

- `targetClass` (mandatory): Defines the name of the in-memory class that this rule is applied to.
- `target` (mandatory): A semicolon-separated list of target class data member names that this rule is potentially updating.
- `embed`: if `true` (the default), the rule is written to the output file if an object of this class is serialized.
- `include`: A comma-separated list of header files that need to be included for the code snippet.
- `code`: The C++ code snippet implementing the rule's actions.

The C++ code snippet has access to the following pre-defined variables:

   * `newObj`: variable pointing to the target in-memory object.
   * `oldObj`: a variable of type {% include ref class="TVirtualObject" %}, behaving as a pointer to the source object.
   * variables representing the data members of the target object declared in the `target` property of the rule.
   * `onfile.variable_name`: variables declared in the source property of the rule

#### Specifying I/O customization rules through the C++ API

> The schema evolution C++ API consists of the follwing classes:

Reviewer 3: ```suggestion
The schema evolution C++ API consists of the following classes:
```

- [TSchemaRule](https://root.cern/doc/master/classROOT_1_1TSchemaRule.html){:target="_blank"}: it represent the rules and their fields have exactly the same meaning as the ones of rules specified in the dictionaries (see [above](#specifying-io-customization-rules-in-a-linkdef-file)).

#### Schema evolution with custom streamers

If you have written your own `Streamer` as described in [Custom streamers](#custom-streamers), you will have to manually add code for each version and manage the evolution of your class. When you add or remove data members, you must modify the `Streamer` by hand. ROOT assumes that you have increased the class version number in the `ClassDef` statement and introduced the relevant test in the read part of the Streamer. For example, if a new version of the `Event` class above includes a new member: `Int_t fNew` the `ClassDef` statement should be changed to `ClassDef(Event,2)` and the following lines should be added to the read part of the `Streamer`:

{% highlight C++ %}

>    if (R__v > 1) R__b >> fNew;

Reviewer 3: I guess `R__b` is the `TBuffer` argument, but do we refer to `R__v` anywhere else?

Author: we don't say what `R__b` is, nor `R__v` (the version number). Rather than just _saying_ it I'd like to point to some reference or some code that shows what signature you can expect a streamer function to have. @pcanal help? :smile:

Reviewer: For R__b, this is the argument to the `Streamer` function, usually declared by `ClassDef`:
```
virtual_keyword void Streamer(TBuffer &R__b) ...
```
R__v is a local variable usually assigned via:
```
   if (b.IsReading()) {
      UInt_t R__s, R__c;
      Version_t R__v = b.ReadVersion(&R__s, &R__c);
      ...... actual custom streamer (often delegating to the default streamer)
      b.CheckByteCount(R__s, R__c, TTree::IsA());
   }
```

>    if (R__v > 1) R__b >> fNew;
>    else fNew = 0;        // set to some default value

Reviewer 3: @pcanal That should be it, no?
```suggestion
   // `buf` below is the TBuffer& argument to the `Streamer` function
   UInt_t R__s, R__c;
   Version_t R__v = buf.ReadVersion(&R__s, &R__c);

   Event::Class()->ReadBuffer(buf, this);
   if (R__v > 1) buf >> fNew;
   else fNew = 0;        // set to some default value

   buf.CheckByteCount(R__s, R__c, Event::IsA());
```

Reviewer: humm .. looks odd.

Reviewer: Right, the odd part is the mix of 'hand coded streamer and automatic streamer.  I would simply replace
```
Event::Class()->ReadBuffer(buf, this);
```
with
```
...
```

Also note that `TClass::ReadBuffer(TBuffer&, void*)` will read the class version and check the byte count (i.e. fail in this case, since the buffer already advanced passed the version).

Reviewer 3: > Right, the odd part is the mix of 'hand coded streamer and automatic streamer. I would simply replace
> 
> ```
> Event::Class()->ReadBuffer(buf, this);
> ```
> 
> with
> 
> ```
> ...
> ```

I have edited the suggestion and added a more explicit comment.  If there are no other suggestions, we could merge this PR.

Reviewer: Go ahead.  It looks good.
